azure/services/async/helper.go

```diff
@@ -0,0 +1,123 @@
+		return azure.WithTransientError(azure.NewOperationNotDoneError(future), reconciler.DefaultReconcilerRequeue)
+	}
+
+	// Resource has been created/deleted/updated.
+	scope.V(2).Info("long running operation has completed", "service", serviceName, "resource", resourceName)
+	scope.DeleteLongRunningOperationState(resourceName, serviceName)
+	return nil
+}
+
+// CreateResource implements the logic for creating a resource Asynchronously.
+func CreateResource(ctx context.Context, scope FutureScope, client Creator, spec azure.ResourceSpecGetter, serviceName string) error {
+	resourceName := spec.ResourceName()
+	rgName := spec.ResourceGroupName()
+
+	// Check if there is an ongoing long running operation.
+	future := scope.GetLongRunningOperationState(resourceName, serviceName)
+	if future != nil {
+		return ProcessOngoingOperation(ctx, scope, client, resourceName, serviceName)
+	}
+
+	// No long running operation is active, so create the resource.
+	scope.V(2).Info("creating resource", "service", serviceName, "resource", resourceName, "resourceGroup", rgName)
+	sdkFuture, err := client.CreateOrUpdateAsync(ctx, spec)
+	if err != nil {
+		if sdkFuture != nil {
+			future, err := converters.SDKToFuture(sdkFuture, infrav1.PutFuture, serviceName, resourceName, rgName)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create resource %s/%s (service: %s)", rgName, resourceName, serviceName)
+			}
+			scope.SetLongRunningOperationState(future)
+			return azure.WithTransientError(azure.NewOperationNotDoneError(future), reconciler.DefaultReconcilerRequeue)
+		}
+
+		return errors.Wrapf(err, "failed to create resource %s/%s (service: %s)", rgName, resourceName, serviceName)
+	}
+
+	scope.V(2).Info("successfully created resource", "service", serviceName, "resource", resourceName, "resourceGroup", rgName)
+	return nil
+}
+
+// DeleteResource implements the logic for deleting a resource Asynchronously.
+func DeleteResource(ctx context.Context, scope FutureScope, client Deleter, spec azure.ResourceSpecGetter, serviceName string) error {
+	resourceName := spec.ResourceName()
+	rgName := spec.ResourceGroupName()
+
+	// Check if there is an ongoing long running operation.
+	future := scope.GetLongRunningOperationState(resourceName, serviceName)
+	if future != nil {
+		return ProcessOngoingOperation(ctx, scope, client, resourceName, serviceName)
+	}
+
+	// No long running operation is active, so delete the resource.
+	scope.V(2).Info("deleting resource", "service", serviceName, "resource", resourceName, "resourceGroup", rgName)
+	sdkFuture, err := client.DeleteAsync(ctx, spec)
+	if err != nil {
+		if azure.ResourceNotFound(err) {
+			// already deleted
+			return nil
+		} else if sdkFuture != nil {
+			future, err := converters.SDKToFuture(sdkFuture, infrav1.DeleteFuture, serviceName, resourceName, rgName)
+			if err != nil {
+				return errors.Wrapf(err, "failed to delete resource %s/%s (service: %s)", rgName, resourceName, serviceName)
```

Comment:
Consider deleting the future if the conversion returns an error here. Is there a reason why you chose not to?

Reply:
the future hasn't been set yet, it gets set on line 114, so there is nothing to delete